port of tR recorder

jsarabia opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/wycliffeassociates/otter-common/108)
<!-- Reviewable:end -->

Commit:
port of tR recorder

Commit:
Removed unneeded tR code

removes some unneeded files and a print statement

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt:
```diff
@@ -0,0 +1,43 @@
+        var max = Float.MIN_VALUE
+
+        for(sample in accumulator) {
+            max = if(max < sample) sample else max
```
Maybe kotlin has a function to find max and min?

aunger replied:
Not that I've seen.  Can get a liiiiittle faster by doing `if (max < sample) max = sample`

jsarabia replied:
fixed, thanks!

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,112 @@
+        val bb = ByteBuffer.allocate(4)
+        bb.order(ByteOrder.LITTLE_ENDIAN)
+        bb.putInt(totalDataSize.toInt())
+        val raf = RandomAccessFile(wav.file, "rw")
```
`.use`

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,112 @@
+    private val outputStream: OutputStream
+    private lateinit var bos: BufferedOutputStream
+    private var audioDataLength: Int = 0
+    private var buffered = false
```
Make `buffered` a val.  Can probably also just get rid of the buffering enum.  I assume you're using the enum so it's a different parameter type than the boolean `append`, and so either of the optional parameters can be omitted unambiguously; however I think just using named params is a good solution to this.

jsarabia replied:
I agree, 
A lot of this was old java code I had tried to just get the IDE to convert to kotlin, rather than rewriting/refactoring much of it (which is what brought in a lot of some of the oddities like using === )

 Not meaning to say I didn't want to clean it up; but just giving some context lol.

Some of it should get some substantial rewriting though, like the FloatRingBuffer, which mixes too much use case specific oddities into something that is misleadingly advertised as general purpose. =\

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Throws(IOException::class)
+    internal fun updateHeader() {
```
This seems to overlap `WavFile.overwriteHeaderData`.  Do you need both?

jsarabia replied:
Probably not. The wavoutputstream helped make sure that the header was overwriten, but I was still needing it when appending metadata (which was code I had removed when moving it over to otter); but if we add that back in it should probably also be done in a way that just uses the outputstream.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,112 @@
+package org.wycliffeassociates.otter.common.io.wav
```
Package name and directory don't match.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,112 @@
+    private var buffered = false
+
+    init {
+        if (wav.file.length().toInt() === 0) {
```
Referential equality tests on primitives is deprecated.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    fun parseHeader() {
```
private?

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    fun parseHeader() {
+        if (file != null && file.length() >= HEADER_SIZE) {
```
`file` isn't nullable

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/ActiveRecordingRenderer.kt:
```diff
@@ -0,0 +1,35 @@
+package org.wycliffeassociates.otter.common.app
```
package doesn't match directory

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/collections/FloatRingBuffer.kt:
```diff
@@ -0,0 +1,63 @@
+                returnBuffer[i * 4] = i.toFloat()
+                returnBuffer[i * 4 + 1] = get(i)
+                returnBuffer[i * 4 + 2] = i.toFloat()
+                returnBuffer[i * 4 + 3] = get(i + 1)
```
I contend that `returnBuffer` only needs to be twice the size of `buffer`.

I think all the `i*4` should be `i*2`.  To underscore the issue, walk through this code (starting with i=0) and see what is stored in `returnBuffer[4]`. Remember `i` is stepping by `+=2`.

Really, though, I content that `returnBuffer` only needs to be the same size as `buffer`.

I don't think the `i.toFloat()` values need to be stored.  The only place this array is used, that I can see, is in [WaveFormLayer](https://github.com/WycliffeAssociates/otter-jvm/pull/463/files#diff-d4bf0e3215a386f96af4ab2cec562955R22), which can just reconstruct these index values on the fly.  If you do that, then you don't need to do this looping at all, and can use two `System.arrayCopy` or [`Array.copyInto`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/copy-into.html)s instead, which should be much faster:

```
    System.arraycopy(buffer, head, returnBuffer, 0, buffer.size - head)
    System.arraycopy(buffer, 0, returnBuffer, buffer.size - head, head)
```

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt:
```diff
@@ -0,0 +1,43 @@
+    var index = 0
+
+    fun add(data: FloatArray) {
+        for(sample in data) {
```
[Space](https://kotlinlang.org/docs/reference/coding-conventions.html#horizontal-whitespace) after `if` and `for`.  Multiple instances throughout the PR.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    fun add(data: Float) {
+        if(index >= accumulator.size){
```
Space before `{` (and after `if`).  Crtl-Alt-L will fix all these in IDEA.

jsarabia replied:
realized there is an extra detail- the kotlin coding convention has to be selected in the code formatting settings (and maybe isn't by default?)

aunger replied:
Could be.  Might also depend on when you initially installed IDEA, because the coding conventions have been updated in the last couple of years.

Commit:
Optimize min/max

Only uses assignment if a new min or max is found

Commit:
Code review corrections

Fixed incorrect package names
Fixed style issues
Buffered enum changed to named parameters
Removed overwriteheaderdata (unused)
changed to .use for file writing

Commit:
Make FloatRingBuffer general purpose

Changes the FloatRingBuffer to not have the return array be domain specific to drawing lines.

As such, the allocation of this array needs to be double the width in the active recording renderer.

Changed the compression value to reflect a bug that sort of skipped over half the array at draw time. This should still be calculated based on the screen width and desired number of seconds on the screen.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt:
```diff
@@ -0,0 +1,43 @@
+    var index = 0
+
+    fun add(data: FloatArray) {
+        for (sample in data) {
```
If you want to avoid this for loop as well as the if statement after it, you could rewrite this function like this:
```
fun add(data: FloatArray) {
        var dataIdx = 0
        while (dataIdx < data.size) {
            val amountUntilFull = accumulator.size - index
            val amountOfDataLeft = data.size - dataIdx 
            val amountToCopy = Math.min(amountUntilFull, amountOfDataLeft)
            
            System.arraycopy(data, dataIdx, accumulator, index, amountToCopy)
            
            index += amountToCopy
            dataIdx += amountToCopy
            
            if (index == accumulator.size) {
                sendDataToRingBuffer()
                index = 0
            }
        }
}
```

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    fun add(data: Float) {
```
Can also rewrite like above

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/PCMCompressor.kt:
```diff
@@ -0,0 +1,43 @@
+        var max = Float.MIN_VALUE
+
+        for (sample in accumulator) {
+            if (max < sample) max = sample
```
Can just use `accumulator.max()` and `.min()`

jsarabia replied:
I could.
Those return nullable types, so I'd have to handle that.
```
/**
 * Returns the smallest element or `null` if there are no elements.
 * 
 * If any of elements is `NaN` returns `NaN`.
 */
public fun FloatArray.min(): Float? {
    if (isEmpty()) return null
    var min = this[0]
    if (min.isNaN()) return min
    for (i in 1..lastIndex) {
        val e = this[i]
        if (e.isNaN()) return e
        if (min > e) min = e
    }
    return min
}
```

This also means I'd be doing two loops, instead of just one.

aunger replied:
True.  For some unjustified reason (probably by false analogy to `System.arraycopy` and Numpy), I often assume the compiler can do better than plain looped java bytecode for primitive arrays.

KJoslyn replied:
Interesting points. And yes, you'd have to handle nulls and do two loops, so I see why you'd want to keep it this way.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/RecordingTimer.kt:
```diff
@@ -0,0 +1,34 @@
+    val timeElapsed: Long
+        get() {
+            val elapsed = System.currentTimeMillis() - startTime + timeStored
+            return if (paused) {
```
In other places, sometimes we've put `if` statements like this all on one line. Do you prefer this way of formatting? I don't think the Kotlin Coding Conventions really state a preference one way or the other.

jsarabia replied:
if I'm honest, I always prefer braces with if/else.
If the case statement needs expanding for some reason, it already has braces. If it needs debugging, it's easier to set the breakpoint.

KJoslyn replied:
Makes sense. I used to think it looked cleaner having the braces too, although now I like conserving space. But the points you make are why sometimes I still opt for braces.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,98 @@
+    fun finishWrite(totalAudioLength: Int) {
+        this.totalAudioLength = totalAudioLength
+    }
+
```
Extra space

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,98 @@
+        header.order(ByteOrder.LITTLE_ENDIAN)
+        header.putInt(16)
+        header.putShort(1) // format = 1
+        header.putShort(NUM_CHANNELS.toShort()) //number of channels
```
Space after //.  (This probably doesn't need to be a comment anyway, the code is self-explanatory)

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,98 @@
+
+        header.order(ByteOrder.LITTLE_ENDIAN)
+        header.putInt(16)
+        header.putShort(1) // format = 1
```
What does a format of 1 mean?

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private fun generateHeaderArray(): ByteArray {
```
The numbers that are hardcoded in this function could maybe use some explanation or a constant. I assume that the use of 8 is byte size- should 44 just be HEADER_SIZE?

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavFile.kt:
```diff
@@ -0,0 +1,98 @@
+ */
+class WavFile(internal val file: File) {
+
+    companion object {
```
There are several unused constants here as of now- ignore if they're meant to be used later

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,108 @@
+class WavOutputStream @Throws(FileNotFoundException::class)
+@JvmOverloads constructor(
+    internal val wav: WavFile,
+    private val append: Boolean = false,
```
Minor: doesn't need to be a val, for now

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,108 @@
+
+    @Throws(IOException::class)
+    internal fun updateHeader() {
+        val totalDataSize = wav.file.length() - 36
```
If this function is kept, should you name these hardcoded values? There are a few in this function.

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/common/collections/FloatRingBuffer.kt:
```diff
@@ -0,0 +1,64 @@
+    private var head = 0
+    private var tail = 0
+    private val buffer = FloatArray(capacity)
+    //this buffer contains x and y values for the high and low in each sample
```
Space after //.  There are other instances of this in this file.

jsarabia committed:
Calculate compression based on screen size/time

Calculates how many samples to compress into one in order to draw the desired number of seconds on the screen.

Also fixes style issues, adds some names to constants

Corrects an issue in the total data size which was previously incorrectly subtracting 36 (probably from removing previous parameters that may have worked out differently when metadata was part of this?)

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/WavFileWriter.kt:
```diff
@@ -0,0 +1,28 @@
+            onComplete()
+        }
+    ).subscribeOn(Schedulers.io())
+        .subscribe()
```
https://kotlinlang.org/docs/reference/coding-conventions.html#chained-call-wrapping
```kotlin
val disposable = Observable
    .using(
        /// ...
    )
    .subscribeOn()
    .subscribe()
```

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/recorder/WavFileWriter.kt:
```diff
@@ -0,0 +1,28 @@
+    private val audioStream: Observable<ByteArray>,
+    private val onComplete: () -> Unit
+) {
+    val writer = Observable.using(
```
Variable doesn't hold a writer.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/wav/WavOutputStream.kt:
```diff
@@ -0,0 +1,111 @@
+) : OutputStream(), Closeable, AutoCloseable {
+
+    private val outputStream: OutputStream
+    private lateinit var bos: BufferedOutputStream
```
If you add an `else` clause to its initialization (line 48), you don't need lateinit and this can be 
```kotlin
private val bos: BufferedOutputStream?
```
